hathor/event/event_manager.py

```diff
@@ -0,0 +1,93 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+from structlog import get_logger
+
+from hathor.event.base_event import BaseEvent
+from hathor.event.storage import EventStorage
+from hathor.pubsub import EventArguments, HathorEvents, PubSubManager
+from hathor.util import Reactor
+
+logger = get_logger()
+
+_GROUP_START_EVENTS = {
+    HathorEvents.REORG_STARTED,
+}
+
+_GROUP_END_EVENTS = {
+    HathorEvents.REORG_FINISHED,
+}
+
+
+class EventManager:
+    def __init__(self, event_storage: EventStorage, reactor: Reactor, peer_id: str):
+        self.log = logger.new()
+        self.clock = reactor
+        self.event_storage = event_storage
+        last_event = event_storage.get_last_event()
+        last_event_type = HathorEvents(last_event.type) if last_event is not None else None
+        # XXX: we must check that the last event either does not belong to an event group or that it just closed an
+        #      event group, because we cannot resume an open group of events that wasn't properly closed before exit
+        assert (
+            last_event is None or
+            last_event.group_id is None or
+            last_event_type in _GROUP_END_EVENTS
+        ), 'an unclosed event group was detected, which indicates the node crashed, cannot resume'
+        self._next_event_id = 0 if last_event is None else last_event.id + 1
+        last_group_id = event_storage.get_last_group_id()
+        self._next_group_id = 0 if last_group_id is None else last_group_id + 1
+        self._current_group_id: Optional[int] = None
+        self._peer_id = peer_id
+
+    def subscribe(self, pubsub: PubSubManager) -> None:
+        """ Subscribe to defined events for the pubsub received
+        """
+        self.log.error(pubsub)
+        events = [
+            HathorEvents.NETWORK_NEW_TX_ACCEPTED,
+            HathorEvents.NETWORK_NEW_TX_VOIDED,
+            HathorEvents.NETWORK_BEST_BLOCK_FOUND,
+            HathorEvents.NETWORK_ORPHAN_BLOCK_FOUND,
+            HathorEvents.LOAD_STARTED,
+            HathorEvents.LOAD_FINISHED,
+            HathorEvents.REORG_STARTED,
+            HathorEvents.REORG_FINISHED,
+            HathorEvents.TX_METADATA_CHANGED,
+            HathorEvents.BLOCK_METADATA_CHANGED,
+        ]
+
+        for event in events:
+            pubsub.subscribe(event, self._persist_event)
+
+    def _persist_event(self, event: HathorEvents, args: EventArguments) -> None:
+        event_data = args.__dict__['event']
+        if event in _GROUP_START_EVENTS:
+            assert self._current_group_id is None, 'cannot start an event group before the last one is ended'
+            self._current_group_id = self._next_group_id
+            self._next_group_id += 1
```

Comment:
Isn't it better to increase this only after the call to `self.event_storage.save_event`?

If an exception happens when running it, we don't want to increase the ids, right?

Reply:
Yeah, that makes sense.